native/xqlitenif/src/r2d2.rs

```diff
@@ -0,0 +1,193 @@
+// This will be the "handle" to an SQLite connection that we return to the Elixir code.
+// It's also the key in the global SQLite connection map.
+static NEXT_ID: AtomicU64 = AtomicU64::new(0);
+
+fn pools() -> &'static XqlitePools {
+    POOLS.get_or_init(|| DashMap::<u64, XqlitePool>::with_capacity(16))
+}
+
+fn create_pool(path: &str) -> Result<u64, r2d2::Error> {
+    let manager = SqliteConnectionManager::file(path);
+    let pool = Pool::new(manager)?;
+    let id = NEXT_ID.fetch_add(1, Ordering::SeqCst);
+    pools().insert(id, pool);
+    Ok(id)
+}
+
+fn get_pool(id: u64) -> Option<XqlitePool> {
+    // This could theoretically deadlock but we never hold onto mutable references to the map.
+    pools().get(&id).map(|x| x.value().clone())
+}
+
+fn remove_pool(id: u64) -> bool {
+    pools().remove(&id).is_some()
+}
+
+#[rustler::nif(schedule = "DirtyIo")]
+fn nif_open<'a>(path: String) -> Result<ResourceArc<XqliteConn>, XqliteError<'a>> {
+    create_pool(&path).map_or_else(
+        |e| Err(XqliteError::CannotOpenDatabase(path, e)),
+        |id| Ok(ResourceArc::new(XqliteConn(id))),
+    )
+}
+
+#[rustler::nif(schedule = "DirtyIo")]
+fn nif_exec<'a>(
+    handle: ResourceArc<XqliteConn>,
+    sql: String,
+) -> Result<Vec<Vec<Result<XqliteVal, XqliteError<'a>>>>, XqliteError<'a>> {
+    let pool = match get_pool(handle.0) {
+        Some(pool) => pool,
+        None => return Err(XqliteError::ConnectionNotFound(*handle)),
+    };
+
+    let conn = match pool.get() {
+        Ok(conn) => conn,
+        Err(e) => return Err(XqliteError::Timeout(e)),
+    };
+
+    let mut stmt = match conn.prepare(&sql) {
+        Ok(stmt) => stmt,
+        Err(e) => return Err(XqliteError::CannotPrepareStatement(sql, e)),
+    };
+
+    let column_count = stmt.column_count();
+    let mut results: Vec<Vec<Result<XqliteVal, XqliteError>>> = Vec::new();
+
+    let rows = match stmt.query_map([], |row| {
+        let mut row_values: Vec<Result<XqliteVal, XqliteError>> =
+            Vec::with_capacity(column_count);
+        for i in 0..column_count {
+            let value: Result<XqliteVal, XqliteError> = match row.get(i) {
+                Ok(val) => Ok(XqliteVal(val)),
+                Err(e) => Err(XqliteError::CannotEncodeValue(i, e)),
+            };
+            row_values.push(value);
+        }
+        Ok(row_values)
+    }) {
+        Ok(mapped_rows) => mapped_rows,
+        Err(err) => return Err(XqliteError::CannotExecute(err)),
+    };
+
+    for row_result in rows {
+        match row_result {
+            Ok(row) => results.push(row),
+            Err(e) => return Err(XqliteError::CannotFetchRow(e)),
+        }
+    }
+
+    Ok(results)
+}
+
+#[rustler::nif(schedule = "DirtyIo")]
+fn nif_close(env: Env, handle: ResourceArc<XqliteConn>) -> Term {
+    if remove_pool(handle.0) {
+        ok().encode(env)
+    } else {
+        (error(), (XqliteError::ConnectionNotFound(*handle))).encode(env)
+    }
+}
```

Comment:
I have used this because this allows me to return only `:ok` to the Elixir code. Returning `Ok(())` results in `{:ok, {}}` being returned to Elixir.

@filmor Is there a better way of doing this without resorting to the `env: Env` variants of the NIF function signature?

Reply:
You know what, nevermind that. I made a huge refactoring here -- https://github.com/dimitarvp/xqlite/pull/9/commits/b17b7c6653b47f05904bffb58af68df327dceeac -- and I'll ask for other feedback Later™. Disregard my question about the better way to just return `:ok` to Elixir.